subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ContinuousBuildActionExecutor.java

```diff
@@ -139,37 +140,46 @@ private void waitForDeployments(BuildAction action, BuildRequestMetaData request
+                    new SingleFirePendingChangesListener(pendingChangesListener),
+                    cancellationToken,
+                    continuousExecutionGate
+                );
+                try {
+                    listenerManager.addListener(fileSystemChangeListener);
+                    lastResult = executeBuildAndAccumulateInputs(action, fileSystemChangeListener, buildSession);
+
+                    if (!fileSystemChangeListener.hasAnyInputs()) {
+                        logger.println().withStyle(StyledTextOutput.Style.Failure).println("Exiting continuous build as no executed tasks declared file system inputs.");
+                        return lastResult;
+                    } else {
+                        cancellableOperationManager.monitorInput(operationToken -> {
+                            fileSystemChangeListener.wait(
+                                () -> logger.println().println("Waiting for changes to input files of tasks..." + determineExitHint(requestContext))
+                            );
+                            if (!operationToken.isCancellationRequested()) {
+                                fileSystemChangeListener.reportChanges(logger);
+                            }
+                        });
+                    }
+                } finally {
+                    listenerManager.removeListener(fileSystemChangeListener);
+                }
 
-                if (!waiter.isWatching()) {
-                    logger.println().withStyle(StyledTextOutput.Style.Failure).println("Exiting continuous build as no executed tasks declared file system inputs.");
-                    return lastResult;
+                if (cancellationToken.isCancellationRequested()) {
+                    break;
                 } else {
-                    cancellableOperationManager.monitorInput(operationToken -> {
-                        FileWatcherEventListener reporter = new DefaultFileWatcherEventListener();
-                        waiter.wait(
-                            () -> logger.println().println("Waiting for changes to input files of tasks..." + determineExitHint(requestContext)),
-                            reporter
-                        );
-                        if (!operationToken.isCancellationRequested()) {
-                            reporter.reportChanges(logger);
-                        }
-                    });
+                    logger.println("Change detected, executing build...").println();
+                    resetBuildStartedTime();
                 }
-            } finally {
-                waiter.stop();
-            }
-
-            if (cancellationToken.isCancellationRequested()) {
-                break;
-            } else {
-                logger.println("Change detected, executing build...").println();
-                resetBuildStartedTime();
             }
+        } finally {
+            virtualFileSystem.registerChangeBroadcaster(null);
```

Comment:
This feels weird. This should either be a setter, or have an `unregisterChangeBroadcaster()` counterpart.